### PR TITLE
#164691214 Implement JWT Token based authentication

### PR DESCRIPTION
**Pull request by reifred**
#### What does this PR do?
* Ensure users receive a JWT token on sign up or log in
#### Description of Task to be completed?
* Changing Authors Haven from session based authentication to token-based authentication
* Users once registered via the registration endpoint, should receive a JWT token to be required on all subsequent calls to all other endpoints that require authentication.
#### How should this be manually tested?
* After cloning this repo, cd into it and switch to this branch
* Set up the virtual environment and activate it. Follow the link below on setting and activating the virtual environment.
https://uoa-eresearch.github.io/eresearch-cookbook/recipe/2014/11/26/python-virtual-env/
* Install the required packages using:
```
pip install -r requirements.txt
```
* Start the development server using ```python manage.py runserver```
* Using Postman test the endpoint below
```
**Registration endpoint**
POST /api/users/
*Sample request body:*
{
  "user":{
  	"username": "user",
    "email": "user@gmail.com",
    "password": "Pass1234"
  }
}

*Sample output on success*

{
    "user": {
        "email": "user@gmail.com",
        "username": "user",
        "token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpZCI6NiwidXNuIjoidXNlciIsImV4cCI6MTU1NjE5MzA1N30.p-hwa4ewi85USSi1Bh4kqkmdMm_Z3CwrnSp66E0IksM"
    }
}

````
* Refer to the README file for the additional routes.
#### What are the relevant pivotal tracker stories?
[#164691214](https://www.pivotaltracker.com/story/show/164691214)

### CODE REVIEW

**Comment by richien** [comment]
This looks accurate, thank you for the good work. However, you forgot to make the PT ID a clickable link to the story on Pivotal Tracker. Other than that, thumbs up, good wo.rk

**Review by richien — COMMENTED** [review]
This looks accurate, thank you for the good work. However, you forgot to make the PT ID a clickable link to the story on Pivotal Tracker. Other than that, thumbs up, good work.

**Review by NicholusMuwonge — CHANGES_REQUESTED** [review]
Hey, @reifred  this is excellent work cant say that I understand everybit of your code but for the bits I understand, Its pretty fine work. I would only recommend you install flake8 to follow the pep8 standards and also for the PR instead of PT story change it to PT story link. Otherwise Nice work.

**Comment by reifred** [comment]
> Hey, @reifred this is excellent work cant say that I understand everybit of your code but for the bits I understand, Its pretty fine work. I would only recommend you install flake8 to follow the pep8 standards and also for the PR instead of PT story change it to PT story link. Otherwise Nice work.

**Comment by reifred** [comment]
> Hey, @reifred this is excellent work cant say that I understand everybit of your code but for the bits I understand, Its pretty fine work. I would only recommend you install flake8 to follow the pep8 standards and also for the PR instead of PT story change it to PT story link. Otherwise Nice work.

Thanks for your review @NicholusMuwonge. However, I used autopep8 which ensures pep8 standards. I request you to review again.

**Review by richien — APPROVED** [review]

File: authors/apps/authentication/backends.py [hunk 1]
```diff
@@ -1,10 +1,46 @@
...
+            msg = 'Invalid authentication. Could not decode token.'
+            raise exceptions.AuthenticationFailed(msg)
+
+        try:
+            user = User.objects.get(username=payload['usn'])
+        except User.DoesNotExist:
+            msg = f"user with username {payload['usn']} is not found"
+            raise exceptions.AuthenticationFailed(msg)
+
+        if not user.is_active:
```
Comment (RonKbS) [inline authors/apps/authentication/backends.py:42]: This check would imply that a user already has an account yet the changes introduced here are for registration. Kindly elaborate on this.

File: authors/apps/authentication/models.py [hunk 2]
```diff
@@ -8,11 +8,12 @@
 )
 from django.db import models
 
+
```
Comment (RonKbS) [inline authors/apps/authentication/models.py:11]: Please refrain from including unnecessary changes such as this, line 16, 37, 38 and any others unrelated to your PT story.

**Review by RonKbS — CHANGES_REQUESTED** [review]
Good work @reifred.

Kindly implement the requested changes and redo the PR description. Specifically; `How should this be manually tested`. You only need to include instructions on how to test the changes you've introduced and not how to set up the application.

**Comment by reifred** [comment]
> Good work @reifred.
> 
> Kindly implement the requested changes and redo the PR description. Specifically; `How should this be manually tested`. You only need to include instructions on how to test the changes you've introduced and not how to set up the application.

Thanks @RonKbS. This is noted let me work on it.

File: authors/apps/authentication/backends.py (hunk 1 above)
Comment (reifred) [reply authors/apps/authentication/backends.py:42]: This is so because in the Acceptance Criteria of the PT board a scenario was illustrated when a user Logs In. That's why I went ahead to add on that check. So I assumed a user logging in yet they have already been deactivated.
```
Acceptance Criteria
GIVEN user successfully signs up/registers or **logs in**
WHEN user passes appropriate JSON object to the register and or login endpoint
THEN user should receive a JWT as the response
```

File: authors/apps/authentication/models.py (hunk 2 above)
Comment (reifred) [reply authors/apps/authentication/models.py:11]: For the unnecessary changes, this came as a result of reformatting the document to pep8.

**Comment by RonKbS** [comment]
Currently running into this error when I try to register a user as per your instructions. Please look into how it can be resolved.

I also do not see why I need to run tests and migrate in relation to these changes so I recommend making more changes to your PR description
Screenshot 2019-03-27 at 11 15 00

**Comment by reifred** [comment]
> Currently running into this error when I try to register a user as per your instructions. Please look into how it can be resolved.
> 
> I also do not see why I need to run tests and migrate in relation to these changes so I recommend making more changes to your PR description
> Screenshot 2019-03-27 at 11 15 00

Thanks, @RonKbS. I was missing the ending slash in the PR description example. I have corrected it kindly retest again. Previous URL was 

> Currently running into this error when I try to register a user as per your instructions. Please look into how it can be resolved.
> 
> I also do not see why I need to run tests and migrate in relation to these changes so I recommend making more changes to your PR description
> Screenshot 2019-03-27 at 11 15 00

Thanks, @RonKbS. I noticed that I had missed an ending slash when giving the PR description. The URL was supposed to be  **/api/users/** but I instead described **/api/users**. Am sorry for this kindly retest again using **/api/users/**

**Comment by reifred** [comment]
> Currently running into this error when I try to register a user as per your instructions. Please look into how it can be resolved.
> 
> I also do not see why I need to run tests and migrate in relation to these changes so I recommend making more changes to your PR description
> Screenshot 2019-03-27 at 11 15 00

Thanks, @RonKbS. I noticed that I had missed an ending slash when giving the PR description. The URL was supposed to be /api/users/ but I instead described /api/users. Am sorry for this kindly retest again using /api/users/

**Review by NicholusMuwonge — APPROVED** [review]
@reifred . Thanks for the changes made I think now you are good to proceed on. Thanks

File: authors/apps/authentication/models.py (hunk 2 above)
Comment (RonKbS) [reply authors/apps/authentication/models.py:11]: Going forward, ensure that the formatting does not introduce changes such as these. I recommend using pylint instead and not relying on autoformatting.

**Review by RonKbS — APPROVED** [review]
